### DISCUSSION TITLE
What is WASMFS memory backend?

### DISCUSSION THREAD

**Discussion by msqr1** [emscripten-core/emscripten#21553] [Q&A] [answered]
Is the memory backend just MEMFS? If so do I need to do ```wasmfs_create_directory("/", 0777, wasmfs_create_memory_backend()) ``` before I use it? Or do I just use it because it is mounted already to "/" by default? I don't use the FS object from JS, only WASMFS in C++.

**Comment by tlively** [comment, marked as answer]
The memory backend is very similar to MEMFS, but it store file contents in Wasm memory rather than in JS. That makes it more efficient in multithreaded applications because each thread can access the memory independently of the others.

You can depend on the fact that it is mounted to "/" by default, so no need to create and mount it yourself.

**Comment by msqr1** [comment]
I still don't understand, how does it know when to use wasm memory or js memory, I thought this has to be mounted in order for it to know I am using wasm memory, otherwise wouldn't this just use JS from the FS object because I was using "/" without mounting? At the end, should I mount? Or just use it, what is more efficient?  @tlively

**Reply by tlively** [reply]
WasmFS is a completely different file system implementation from the default file system, so it never uses MEMFS directly. If you link with -sWASMFS, then "/" is mounted to the WasmFS memory backend (using Wasm memory) by default and otherwise "/" is mounted to MEMFS (using JS memory) by default.

**Reply by msqr1** [reply]
Thank you very much! So why do we provide a mount option for this? If it is already mounted? If I mount it to another directory this would make 2 wasmfs memfs
@tlively

**Reply by tlively** [reply]
The mount option for the WasmFS memory backend is for niche cases where you want to mount a directory with the memory backend under a directory mounted with a different backend. So you can create a file system that looks like this:

```
'/' (memory backend)
 |
'my_files' (OPFS backend)
 |
'temp_files' (memory backend)
```

**Reply by msqr1** [reply]
Thanks!
